### PR TITLE
Remove image dependency on fex for mainline

### PR DESCRIPTION
**Pull request by jlucius**
Remove image dependency on building fex files which are not needed
any more for mainline kernel (replaced by devicetree).
Add dependency for fex to linux-sunxi kernel.
Also update to latest sunxi-linux revision.

Signed-off-by: Jens Lucius <info@jenslucius.com>

### CODE REVIEW

**Comment by jlucius** [comment]
The patch is currently build time tested with the sunxi-kernel (should not influence mainline kernel builds)

File: classes/sdcard_image-sunxi.bbclass [hunk 1]
```diff
@@ -35,10 +35,9 @@ IMAGE_DEPENDS_sunxi-sdimg += " \
 			dosfstools-native \
 			virtual/kernel \
 			virtual/bootloader \
-                        sunxi-board-fex \
 			"
 
-rootfs[depends] += "virtual/kernel:do_deploy sunxi-board-fex:do_deploy"
+rootfs[depends] += "virtual/kernel:do_deploy"
```
Comment (radolin) [inline classes/sdcard_image-sunxi.bbclass]: I also looked at this during the weekend and I have the suspiction that the above line doesn't actually do anything. I think the correct syntax would be do_rootfs[depends] += .... I'd say it will be okay to remove it completely, as in image.bbclass we have do_build[depends] += "virtual/kernel:do_deploy" which should be enough.

Comment (radolin) [inline classes/sdcard_image-sunxi.bbclass]: I also looked at this during the weekend and I have the suspiction that the above line doesn't actually do anything. I think the correct syntax would be do_rootfs[depends] += .... I'd say it will be okay to remove it completely, as in image.bbclass we have do_build[depends] += "virtual/kernel:do_deploy" which should be enough.

File: recipes-kernel/linux/linux-sunxi_3.4.bb [hunk 2]
```diff
@@ -3,10 +3,11 @@ require linux.inc
 DESCRIPTION = "Linux kernel for Allwinner a10/a20 processors"
 
 COMPATIBLE_MACHINE = "(sun4i|sun5i|sun7i)"
+DEPENDS = "sunxi-board-fex"
```
Comment (radolin) [inline recipes-kernel/linux/linux-sunxi_3.4.bb]: Shouldn't this be RDEPENDS, as the results from bulding the sunxi-board-fex recipe aren't needed for the kernel compilation, but rather at runtime.

File: recipes-kernel/linux/linux-sunxi_3.4.bb [hunk 3]
```diff
@@ -3,10 +3,11 @@ require linux.inc
...
 
-PV = "3.4.103"
+PV = "3.4.104"
 PR = "r1"
```
Comment (radolin) [inline recipes-kernel/linux/linux-sunxi_3.4.bb]: May be put the version bump in a separate commit/pull request? I'm quite new to the meta-sunxi community so I'm not sure what the gudelines are around here.

File: recipes-kernel/linux/linux-sunxi_3.4.bb (hunk 2 above)
Comment (jlucius) [inline recipes-kernel/linux/linux-sunxi_3.4.bb]: In that case the end result is the same, but it might be better to use RDEPENDS, but that depends ... :-) Can change that if necessary

File: recipes-kernel/linux/linux-sunxi_3.4.bb (hunk 3 above)
Comment (jlucius) [inline recipes-kernel/linux/linux-sunxi_3.4.bb]: It would be better to put it in it´s own pull request, but the change is not that big so I used the same commit.

**Comment by ebutera** [comment]
ack from me, just change depends to rdepends.
The version bump should have been in a separate commit but it's ok for me.

i would wait for an opinion from @soderstrom-rikard since he worked on this too, i think this solution is simpler and cleaner but let's hear from him too.
And of course from @naguirre too

**Comment by soderstrom-rikard** [comment]
I totally agree with @ebutera that this solution is the cleanest. It has my vote.

I also favour `RDEPENDS` before `DEPENDS` even if the end result in this case is the same. It will be  clearer for anyone looking at this again in the future.

**Comment by jlucius** [comment]
Ok will change that afte work today as well.

**Comment by radolin** [comment]
Also, I think it should be RDEPENDS_kernel-base, not just RDEPENDS.

**Comment by jlucius** [comment]
My choice would have been RDEPENDS_${PN}, why depend on kernel-base?

**Comment by radolin** [comment]
I don't think RDEPENDS_${PN} would work, as that would be equivalent to RDEPENDS_linux-sunxi and there is no such package named linux-sunxi. It's true that for a lot of recipes the recipe and package name are the same, or at least one of the package names is the same as the recipe name, as one recipe can provide multiple packages. But that's not a requirement and in the case of linux kernel recipes it's not true.

You can see that the linux-sunxi_3.4.bb recipe includes linux.inc, which inherits the kernel.bbclass.
Inside the kernel.bbclass we have:

```
# kernel-base becomes kernel-${KERNEL_VERSION}
# kernel-image becomes kernel-image-${KERNEL_VERISON}
PACKAGES = "kernel kernel-base kernel-vmlinux kernel-image kernel-dev kernel-modules"
FILES_${PN} = ""
FILES_kernel-base = "/lib/modules/${KERNEL_VERSION}/modules.order /lib/modules/${KERNEL_VERSION}/modules.builtin"
FILES_kernel-image = "/boot/${KERNEL_IMAGETYPE}*"
...
```
So in fact the files to include in the default package name are set to nothing, instead the kernel-base and kernel-mage packages are created (which later get renamed to include the version number, like kernel-3.4.90-...). The reason to depend on kernel-base and not kernel-image is that it's possible to exclude kernel-image from your rootfs if you want to save some space for example.

You can check that's what is used for the device tree files in mainline kernel recipes, and for the sunxi kernel FEX/bin files are exactly the equivalent of devicetree.

**Comment by jlucius** [comment]
Actually it works because in this case there are no files that needs to be included by the build process, but the sunxi-sdcard image copies the files, but I agree with you that your suggestion is the best way to go. Will change it an test it. Everbody agrees?

**Comment by soderstrom-rikard** [comment]
@jlucius  :+1:

**Comment by jlucius** [comment]
Well, Ok, good idea, but this time the build throws an error:
Computing transaction...error: Can't install kernel-3.4.104-3.4.104-r1@olinuxino_a20lime: no package provides sunxi-board-fex

the sunxi-board-fex does not provide any package (in recipe: PACKAGES = "") and no files are installed into the rootfs, so this is the difference to the device tree, where the device-tree files are installed into the rootfs.

So either go back to RDEPENDS_${PN} (which works), go to DEPENDS or change the sunxi-board-fex recipe. Any suggestions?

**Comment by radolin** [comment]
Heh, interesting! I hadn't noticed that the sunxi-board-fex recipe produces no packages. 

If RDEPENDS_${PN} works I see no problem to use that, I just don't understand the exact mechanism how it works. If someone can explain I'm curious to know, or I'll play with it if I have the time.

Otherwise we could try to make sunxi-board-fex produce some mpty/dummy package, but that would be more changes.

**Comment by jlucius** [comment]
Before I start messing with the sunxi-board-fex, I´ll just use RDEPENDS_${PN}. Fex files are just for compatibility with the old sunxi-kernel, so I rather would not try to optimize the last bit out of it.

Pull is updated

**Comment by jlucius** [comment]
Ok and btw: build and runtime tested

**Comment by ebutera** [comment]
I agree that since it works and it's for the old kernel there is no need to waste time on it, so i'm merging it.

**Comment by naguirre** [comment]
Thanks for your work., i'm still using the old 3.4 kernel, so i will try a fresh build with this PR to see if everything is ok on my side.
